### PR TITLE
 [#392] expose parent_id, parent_name and resc_hierarchy for resource objects

### PR DESCRIPTION
**Pull request by d-w-moore**
Main motivation is avoid ambiguity: new properties `resc.parent_id` and `resc.parent_name` are explicit in what they provide, whereas `resc.parent` (an already existing iRODSResource attribute) is just what Genquery returns in RESC_PARENT (that is, the parent's name on iRODS <= 4.1 and parent's ID on iRODS >=4.2) See #392

### CODE REVIEW

File: irods/resource.py [hunk 1]
```diff
@@ -38,6 +38,54 @@ def __init__(self, manager, result=None):
...
+            if sess.server_version < (4, 2, 0):
+                self._parent_name = self.parent
+            else:
+                self._parent_name = sess.query(Resource).filter(Resource.id == self.parent).one()[Resource.name]
+        return self._parent_name
+
+    ## Cached property to expose resource hierarchy string
+
+    @property
+    def resc_hierarchy(self):
```
Comment (trel) [inline irods/resource.py:72]: not just ... hierarchy?    this is a property of a resource... so already has that context?

Comment (d-w-moore) [reply irods/resource.py:72]: or `hierachy_string` maybe.

**Comment by trel** [comment]
seems good to me so far.

File: print_resc_properties.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import irods.test.helpers as h
+import sys
+
+s = h.make_session()
+
+r = s.resources.get(sys.argv[1])
+
+print('id = ',r.parent_id)
+print('name = ',r.parent_name)
+print('resc_hier = ',r.hierarchy)
```
Comment (trel) [inline print_resc_properties.py:12]: when you write the test - please confirm the string-ness and int-ness of the return types.

File: irods/resource.py (hunk 1 above)
Comment (d-w-moore) [reply irods/resource.py]: agreed, it will just be `hierarchy`.

Comment (d-w-moore) [reply irods/resource.py]: Cannot resolve the conversation however, maybe because it's in draft status.

File: print_resc_properties.py (hunk 2 above)
Comment (d-w-moore) [reply print_resc_properties.py:12]: will do!

Comment (d-w-moore) [reply print_resc_properties.py:12]: @trel good call. it did seem under some circumstances the id returned in the result set was a string

File: irods/test/resource_test.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    from helpers import create_simple_resc_hierarchy
+
+    def setUp(self):
+        self.sess = helpers.make_session()
+
+    def test_new_resource_properties_for_parent_and_hierarchy__392(self):
+        ses = self.sess
+        root = 'pt_392'
+        leaf = 'leaf_392'
+        with self.create_simple_resc_hierarchy(root,leaf) as resc_hier:
```
Comment (trel) [inline irods/test/resource_test.py:21]: please make the test at least three levels... rather than just parent/child.

**Comment by d-w-moore** [comment]
Officially ready for review

File: irods/test/resource_test.py (hunk 3 above)
Comment (d-w-moore) [reply irods/test/resource_test.py:21]: Oh... like deferred/pt/leaf

Comment (trel) [reply irods/test/resource_test.py:21]: sure, or even just pt/pt/leaf - mostly that we squeeze out any corner cases and assumptions.

Comment (d-w-moore) [reply irods/test/resource_test.py:65]: Ok, done... I think.

File: irods/test/resource_test.py [hunk 4]
```diff
@@ -14,6 +14,50 @@ class TestResource(unittest.TestCase):
...
+                        parent_resc = None
+                        hier_str = root
+                        # Assert that the hierarchy and parent properties hold at each level, in both tree branches.
+                        for n,resc in enumerate(hierarchy):
+                            if n > 0:
+                                hier_str += ";{}".format(resc.name)
+                            self.assertEqual( resc.parent_id, (None if n == 0 else parent_resc.id))
+                            self.assertEqual( resc.parent_name, (None if n == 0 else parent_resc.name))
+                            self.assertEqual( resc.hierarchy, hier_str)
+                            self.assertIs( type(resc.hierarchy), str )                   # type of hierarchy field is string.
```
Comment (d-w-moore) [inline irods/test/resource_test.py:46]: Should be assertEq for hier string.

File: irods/test/resource_test.py [hunk 5]
```diff
@@ -14,6 +14,50 @@ class TestResource(unittest.TestCase):
...
+
+    def test_resource_properties_for_parent_and_hierarchy_at_3_levels__392(self):
+        ses = self.sess
+        root = 'deferred_392'
+        pt = 'pt_392'
+        leaf = 'leaf_392'
+        root_resc = ses.resources.create(root,'deferred')
+        try:
+            # Create two (passthru + storage) hierarchies below the root: ie. pt0;leaf0 and pt1;leaf1
+            with self.create_simple_resc_hierarchy(pt + "_0",leaf + '_0') as resc_hier_0, \
```
Comment (d-w-moore) [inline irods/test/resource_test.py:27]: Maybe more consistent wrote usage.

File: irods/resource.py [hunk 6]
```diff
@@ -38,6 +41,54 @@ def __init__(self, manager, result=None):
...
+
+    @property
+    def hierarchy(self):
+        if self._hierarchy == '':
+            self._hierarchy = ';'.join(r.name for r in reversed(self.get_parent_chain()))
+        return self._hierarchy
+
+    ## Retrieve chain of parent objects to top level parent
+
+    def get_parent_chain(self):
```
Comment (alanking) [inline irods/resource.py:82]: There is a function in `server/core` called `irods::leaf_id_to_hier` which effectively does the same thing as this function. I personally find that parlance easier to understand than "parent chain". Should we consider a name similar to the other function in the server? Just trying to bring together a common language, I guess...

File: irods/test/resource_test.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                    # Adopt both passthru's as children under the main root (deferred) node.
+                    ses.resources.add_child( root, pt + "_0")
+                    ses.resources.add_child( root, pt + "_1")
+                    # Now we have two different 3-deep hierarchies (root;pt0;leaf0 and root;pt1;leaf) sharing the same root node.
+                    # Descend each and make sure the relationships hold
+                    for mid in root_resc.children:
+                        hierarchy = [root_resc, mid, mid.children[0]]
+                        parent_resc = None
+                        hier_str = root
+                        # Assert that the hierarchy and parent properties hold at each level, in both tree branches.
```
Comment (alanking) [inline irods/test/resource_test.py:39]: Is the idea here that we are testing `.hierarchy` from the root resource, middle resource(s), and leaf resource(s)?

**Review by alanking — COMMENTED** [review]
Overall, looks good to me. Just had a couple of questions

File: irods/test/resource_test.py (hunk 7 above)
Comment (d-w-moore) [reply irods/test/resource_test.py:39]: yes, exactly.

File: irods/resource.py (hunk 6 above)
Comment (d-w-moore) [reply irods/resource.py:82]: In this case, it's not necessarily a leaf, but we could do something in that spirit.  How about `node_to_hier`, then?

Comment (d-w-moore) [reply irods/resource.py:82]: or `node_to_topdown_hier` perhaps? not sure what folks would prefer.

File: irods/resource.py [hunk 8]
```diff
@@ -38,6 +41,54 @@ def __init__(self, manager, result=None):
...
+
+    @property
+    def hierarchy(self):
+        if self._hierarchy == '':
+            self._hierarchy = ';'.join(r.name for r in self.node_to_hier())
+        return self._hierarchy
+
+    ## Retrieve chain of parent objects to top level parent
+
+    def node_to_hier(self):
```
Comment (alanking) [inline irods/resource.py:82]: I have another opinion now that I'm looking at this again...

Is there a reason that this function has to be exposed? Could it just be embedded in `hierarchy`?

Any others have thoughts on this?

**Review by alanking — COMMENTED** [review]
I had another take on `node_to_hier`, please don't get upset :p

File: irods/resource.py (hunk 8 above)
Comment (trel) [reply irods/resource.py:82]: I think I agree - not too happy with the name (yet).   Keeping it internal seems an easy solution for now.

Comment (d-w-moore) [reply irods/resource.py:82]: I think embedding it would be fine, yes.  I'll wait on others to weigh in.

**Comment by d-w-moore** [comment]
> I had another take on `node_to_hier`, please don't get upset :p

oops, I'll have to read my reply again, I guess : / ... i didn't think I was upset,  more just thinking aloud ....
It's just that I lacked understanding, I think. I never know how exhaustive to be with my naming and the majority of my history is much more along lines of a, b, c being acceptable variable names ; )  .  And, there was also a time when C compilers didn't accept identifiers of more than 40 characters, believe it or not ....

File: irods/resource.py (hunk 8 above)
Comment (d-w-moore) [reply irods/resource.py:82]: could we call it _hierarchy_for_node and just imply its hiddenness? this has good precedent.  if you embed it in another function, it causes runtime overhead.

Comment (d-w-moore) [reply irods/resource.py:82]: ...one way in which C++ is definitely Python's superior.

Comment (korydraughn) [reply irods/resource.py:82]: So `node_to_hier` returns the hierarchy as a list of `iRODSResource` objects, correct?

Embedding it is a fine solution. Getting the hierarchy as a list of objects seems like a useful thing too (as a convenience).

How do we feel about using more letters/words?
- `hierarchy` -> `hierarchy_string`
- `node_to_hier` -> `hierarchy_as_list_of_resource_objects` (I feel there is a better name hiding in there)

Comment (alanking) [reply irods/resource.py:82]: "Embedded" perhaps was not the right word. I guess I meant "integrated".

In my opinion, this doesn't need to be a separate method because `hierarchy` is calling it exclusively, and that is the only situation in which this is useful. In fact, I kind of see these as one and the same, which is why I'm suggesting integration.

The primary reason I can see for keeping these separated is for providing multiple options with respect to return types (list versus semicolon-delimited-string). So, should we choose to keep these separated, I think `hierarchy_string` and `hierarchy_as_list_of_resource_objects` definitely conveys what this method is doing very clearly. 

Alternatively, `hierarchy` could be a class (or sub-class) which provides a `__str__()` method to get the string version. Or something.

Comment (d-w-moore) [reply irods/resource.py]: Ok,  I renamed the method, which I think is as much as the present issue demands.  If there's a perceived need for hierarchy to be its own concept (with class and stringization methods), then I can make a new issue.  But I think this fills the bill for now.

Comment (d-w-moore) [reply irods/resource.py]: A last note - I took the present approach - with "B being called exclusively by A" - as being a way to simplify the logical flow of A and make it more readable.

**Comment by trel** [comment]
please resolve and squash to taste.

**Comment by d-w-moore** [comment]
Sure, will squash, but I kind of wanted @alanking to look at what would be (at least now, in my punch-drunk state) the likely result of the Hierarchy class idea.

Although it "seems to work",  I have a really strong distaste for it. https://github.com/d-w-moore/python-irodsclient/tree/392-exper.m

File: irods/resource.py (hunk 8 above)
Comment (alanking) [reply irods/resource.py]: I think the implementation in the latest commit is fine. Let's hold off on the `hierarchy` type idea since this implementation working and clear.

**Review by alanking — COMMENTED** [review]
I think the open conversation can be resolved. Let's squash em down

**Comment by d-w-moore** [comment]
squashed!

**Review by alanking — APPROVED** [review]
Please await one more approval and we can add #s

**Review by korydraughn — APPROVED** [review]
Add the pounds.

**Comment by d-w-moore** [comment]
# added
